upf/dnf.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import upf.environment
+import upf.operators as op
+import upf.walkers as walkers
+from upf.walkers.dag import DagWalker
+from upf.fnode import FNode
+from upf.simplifier import Simplifier
+from typing import List, Dict, Tuple
+from itertools import product
+
+
+class Nnf():
+    """Class used to transform a logic expression into the equivalent
+    Negational Normal Form expression.
+
+    This is done first by removing all the Implications and Equalities,
+    then by pushing all the not to the leaves of the Tree representing the expression."""
+
+    def __init__(self, env: 'upf.environment.Environment'):
+        self.env = env
+        self.manager = env.expression_manager
+
+    def get_nnf_expression(self, expression: FNode) -> FNode:
```

Comment:
Something like this should work and it is a little more clear:
```python
    def get_nnf_expression(self, expression: FNode) -> FNode:
        stack = []
        statuses = []
        stack.append((True, expression))
        statuses.append(False)
        solved = []
        while len(stack) > 0:
            p, e = stack.pop()
            status = statuses.pop()
            if status:
                if e.is_not():
                    pass
                elif e.is_and():
                    args = [solved.pop() for i in range(len(e.args()))]
                    if p:
                        new_e = self.manager.And(args)
                    else:
                        new_e = self.manager.Or(args)
                    solved.append(new_e)
                elif e.is_or():
                    args = [solved.pop() for i in range(len(e.args()))]
                    if p:
                        new_e = self.manager.Or(args)
                    else:
                        new_e = self.manager.And(args)
                    solved.append(new_e)
                else:
                    raise
            else:
                if e.is_not():
                    stack.append((p, e))
                    statuses.append(True)
                    stack.append((not p, e.arg(0)))
                    statuses.append(False)
                elif e.is_and() or e.is_or():
                    stack.append((p, e))
                    statuses.append(True)
                    for arg in e.args():
                        stack.append((p, arg))
                        statuses.append(False)
                elif e.is_implies():
                    stack.append((p, self.manager.Or(self.manager.Not(e.arg(0)), e.arg(1))))
                    statuses.append(False)
                elif e.is_iff():
                    stack.append((p, self.manager.Or(self.manager.And(e.arg(0), e.arg(1)),
                                                     self.manager.And(self.manager.Not(e.arg(0)),
                                                                      self.manager.Not(e.arg(1))))))
                    statuses.append(False)
                else:
                    if p:
                        solved.append(e)
                    else:
                        solved.append(self.manager.Not(e))
        return solved.pop()

```

Reply:
Yes it should work! It is also possible to "skip" iterations for Implies and Iff, by decomposing in place the or, the and and the not, at the cost of readability. Like for Implies:

```suggestion
    def get_nnf_expression(self, expression: FNode) -> FNode:
        ....
            elif e.is_implies():
                    na1 = self.manager.Not(e.arg(0))
                    new_e = self.manager.Or(na1, e.arg(1))
                    stack.append((p, new_e, True))
                    stack.append((not p, e.arg(0), False))
                    stack.append((p, e.arg(1), False))
```

What do you think about that? It is also possible for Iff, based on what you like better we can choose the best option.